grails-app/services/org/pih/warehouse/importer/InventoryImportDataService.groovy

```diff
@@ -128,74 +138,172 @@ class InventoryImportDataService implements ImportDataService {
-            }
+    private InventoryItem parseInventoryItem(Product product, def lotNumberRaw, def expirationDateRaw) {
+        String lotNumber = lotNumberRaw instanceof Double ? lotNumberRaw.toInteger() : lotNumberRaw
 
-            transaction.addToTransactionEntries(transactionEntry)
+        Date expirationDate = null
+        if (expirationDateRaw instanceof String) {
+            expirationDate = EXPIRATION_DATE_FORMAT.parse(expirationDateRaw)
+            Calendar calendar = Calendar.getInstance()
+            calendar.setTime(expirationDate)
+            expirationDate = calendar.getTime()
+        } else if (expirationDateRaw instanceof Date) {
+            expirationDate = expirationDateRaw
+        } else if (expirationDateRaw instanceof LocalDate) {
+            expirationDate = expirationDateRaw.toDate()
+        }
+
+        return inventoryService.findAndUpdateOrCreateInventoryItem(product, lotNumber, expirationDate)
+    }
+
+    private Location parseBinLocation(String binLocationName, Location parentLocation) {
+        if (binLocationName == null) {
+            return null
+        }
+
+        Location binLocation = Location.findByNameAndParentLocation(binLocationName, parentLocation)
+        assert binLocation != null
+        return binLocation
+    }
+
+    /**
+     * Convenience POJO for holding the results of the inventory import.
+     */
+    private class InventoryImportData {
+        Map<String, InventoryImportDataRow> rows = [:]
+        Set<Product> products = []
+
+        void put(Location binLocation, InventoryItem inventoryItem, def rowRaw) {
+            products.add(inventoryItem.product)
+
+            String key = ProductAvailabilityService.constructAvailableItemKey(binLocation, inventoryItem)
+            rows.put(key, new InventoryImportDataRow(
+                    binLocation,
+                    inventoryItem,
+                    rowRaw.quantity.toInteger() as Integer,
+                    rowRaw.comments as String))
+        }
+
+        InventoryImportDataRow get(Location binLocation, InventoryItem inventoryItem) {
+            String key = ProductAvailabilityService.constructAvailableItemKey(binLocation, inventoryItem)
+            return rows.get(key)
+        }
+    }
+
+    private class InventoryImportDataRow {
+        Product product
+        Location binLocation
+        InventoryItem inventoryItem
+        Integer quantity
+        String comments
+
+        InventoryImportDataRow(Location binLocation, InventoryItem inventoryItem, Integer quantity, String comments) {
+            this.product = inventoryItem.product
+            this.binLocation = binLocation
+            this.inventoryItem = inventoryItem
+            this.quantity = quantity
+            this.comments = comments
         }
-        // Force refresh of inventory snapshot table
-        transaction.forceRefresh = Boolean.TRUE
-        transaction.save(flush: true, failOnError: true)
```

Comment:
TODO: investigate if we still need to flush. I changed it to not but I'm not sure why it was originally required

Reply:
It seems to be functioning fine without the forced flush so I'm leaving it out